File: frontend/src/test-fixtures/basicDashboard.ts
import type { BasicDashboardResponse, IndicatorValue, SeriesPoint } from "@/types/basic-dashboard";

export const makeIndicator = (overrides: Partial<IndicatorValue> = {}): IndicatorValue => ({
  key: "gross_loan_portfolio",
  group: "loans",
  value: 1000,
  unit: "currency",
  status: "computed",
  previous: null,
  change_pct: null,
  formula: "outstanding_value_male + outstanding_value_female",
  sources: ["outstanding_value_male", "outstanding_value_female"],
  note: null,
  ...overrides,
});

const point = (
  label: string,
  year: number,
  value: number,
  extra: Record<string, number> = {},
): SeriesPoint => ({
  period_label: label,
  reporting_year: year,
  period_type: "QUARTERLY",
  period_value: label.slice(0, 2),
  values: { total_assets: value, ...extra },
});

export const basicDashboardFixture: BasicDashboardResponse = {
  scope: {
    level: "consolidated",
    cooperatives_reporting: 3,
    cooperatives_in_scope: 4,
    reporting_year: 2026,
    period_type: "QUARTERLY",
    period_value: "Q1",
    period_label: "Q1 2026",
    currency: "USD",
    native_currency: "SZL",
    rate_to_usd: 18.5,
    available_periods: [
      { reporting_year: 2026, period_type: "QUARTERLY", period_value: "Q1", label: "Q1 2026" },
      { reporting_year: 2025, period_type: "QUARTERLY", period_value: "Q4", label: "Q4 2025" },
    ],
  },
  thresholds: { liquidity_minimum_pct: 15, institutional_capital_minimum_pct: 8 },
  indicators: [
    makeIndicator(),
    makeIndicator({
      key: "par_gt_30_pct",
      group: "risk",
      unit: "percent",
      value: 2.94,
      previous: 4.19,
      change_pct: -29.8,
    }),
    makeIndicator({
      key: "par_gt_90_pct",
      group: "risk",
      unit: "percent",
      value: null,
      status: "not_reported",
    }),
    makeIndicator({
      key: "registered_members",
      group: "membership",
      unit: "count",
      value: 6992,
    }),
  ],
  series: {
    asset_evolution: [
      point("Q1 2025", 2025, 3130874),
      point("Q4 2025", 2025, 3348330),
      point("Q1 2026", 2026, 3485818),
    ],
    par_trend: [
      point("Q1 2025", 2025, 0, { par_gt_30_pct: 5.3 }),
      point("Q1 2026", 2026, 0, { par_gt_30_pct: 2.94 }),
    ],
  },
  market_share: {
    by_assets: [
      { cooperative_id: "a", name: "SNAT", value: 100, share_pct: 60 },
      { cooperative_id: "b", name: "Bunye", value: 66, share_pct: 40 },
    ],
    by_loans: [],
  },
  demographics: {
    registered: { male: 100, female: 50 },
    active: { male: 80, female: 40 },
    age: { age_18_25: 10, age_26_35: 20, age_36_60: 30, age_61_plus: 5 },
    board: { male: 3, female: 2 },
    executive: { male: 2, female: 1 },
    credit_committee: { male: 1, female: 1 },
    savings_accounts: { male: 90, female: 45 },
    loan_accounts: { male: 60, female: 30 },
  },
  cooperatives: [
    {
      cooperative_id: "a",
      name: "SNAT",
      region: "Manzini",
      sector: "Finance",
      total_members: 3000,
      total_assets: 1000,
      total_deposits: 700,
      gross_loans: 600,
      par_gt_30_pct: 3,
      liquidity_ratio_pct: 12,
      institutional_capital_ratio_pct: 14,
      net_income: 50,
    },
    {
      cooperative_id: "b",
      name: "Bunye",
      region: null,
      sector: null,
      total_members: 2000,
      total_assets: 660,
      total_deposits: 400,
      gross_loans: 300,
      par_gt_30_pct: null,
      liquidity_ratio_pct: null,
      institutional_capital_ratio_pct: null,
      net_income: -10,
    },
  ],
};

File: backend/src/api/dto/basic_dashboard.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use utoipa::{IntoParams, ToSchema};
use uuid::Uuid;

#[derive(Debug, Clone, Deserialize, IntoParams, ToSchema)]
pub struct BasicDashboardParams {
    pub reporting_year: Option<i32>,
    /// YEARLY | SEMI_ANNUAL | QUARTERLY | MONTHLY
    pub period_type: Option<String>,
    /// e.g. "Q1", "H2", "2026", "03"
    pub period_value: Option<String>,
    pub region: Option<String>,
    pub sector: Option<String>,
    /// Restrict to one cooperative (individual view).
    pub cooperative_id: Option<Uuid>,
    /// `usd` (default, converted with the submission's rate) or `native` (as entered).
    pub currency: Option<String>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct PeriodOption {
    pub reporting_year: i32,
    pub period_type: String,
    pub period_value: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct DashboardScope {
    /// "individual" (one cooperative) or "consolidated" (several).
    pub level: String,
    pub cooperatives_reporting: u32,
    pub cooperatives_in_scope: u32,
    pub reporting_year: Option<i32>,
    pub period_type: Option<String>,
    pub period_value: Option<String>,
    pub period_label: String,
    /// Currency of every monetary value in the response ("USD" or "SZL").
    pub currency: String,
    pub native_currency: String,
    /// Units of native currency per 1 USD when converted, else null.
    pub rate_to_usd: Option<f64>,
    pub available_periods: Vec<PeriodOption>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct DashboardThresholds {
    pub liquidity_minimum_pct: f64,
    pub institutional_capital_minimum_pct: f64,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct IndicatorValue {
    pub key: String,
    /// membership | savings | loans | risk | liquidity | structure | capital | profitability | governance
    pub group: String,
    pub value: Option<f64>,
    /// count | currency | percent | ratio
    pub unit: String,
    /// computed | approximate | not_reported
    pub status: String,
    pub previous: Option<f64>,
    pub change_pct: Option<f64>,
    /// Human-readable formula used for the value.
    pub formula: String,
    /// Questionnaire field keys the value is derived from.
    pub sources: Vec<String>,
    pub note: Option<String>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct SeriesPoint {
    pub period_label: String,
    pub reporting_year: i32,
    pub period_type: String,
    pub period_value: String,
    /// series-specific named values (see `series` docs on the response)
    pub values: HashMap<String, f64>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct ShareRow {
    pub cooperative_id: Uuid,
    pub name: String,
    pub value: f64,
    pub share_pct: f64,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct MarketShare {
    pub by_assets: Vec<ShareRow>,
    pub by_loans: Vec<ShareRow>,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct GenderCount {
    pub male: f64,
    pub female: f64,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct AgeBands {
    pub age_18_25: f64,
    pub age_26_35: f64,
    pub age_36_60: f64,
    pub age_61_plus: f64,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct Demographics {
    pub registered: GenderCount,
    pub active: GenderCount,
    pub age: AgeBands,
    pub board: GenderCount,
    pub executive: GenderCount,
    pub credit_committee: GenderCount,
    pub savings_accounts: GenderCount,
    pub loan_accounts: GenderCount,
}

#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct CooperativeRow {
    pub cooperative_id: Uuid,
    pub name: String,
    pub region: Option<String>,
    pub sector: Option<String>,
    pub total_members: f64,
    pub total_assets: Option<f64>,
    pub total_deposits: Option<f64>,
    pub gross_loans: Option<f64>,
    pub par_gt_30_pct: Option<f64>,
    pub liquidity_ratio_pct: Option<f64>,
    pub institutional_capital_ratio_pct: Option<f64>,
    pub net_income: Option<f64>,
}

/// Response of `GET /api/v1/analytics/basic-dashboard`.
///
/// `series` keys (each a chronological list of points, `values` keys in brackets):
/// - `asset_evolution` (total_assets)
/// - `savings_trend` (total_deposits)
/// - `loan_portfolio` (gross_loans)
/// - `par_trend` (par_gt_30_pct, par_gt_90_pct, portfolio_at_risk_pct)
/// - `liquidity` (maintained_pct, minimum_pct, gap_pct)
/// - `financial_structure` (earning_asset_ratio, member_savings_ratio, member_share_ratio, borrowed_funds_ratio)
/// - `profitability` (net_income)
/// - `institutional_capital` (ratio_pct, minimum_pct, excess_pct)
#[derive(Debug, Clone, Serialize, ToSchema)]
pub struct BasicDashboardResponse {
    pub scope: DashboardScope,
    pub thresholds: DashboardThresholds,
    pub indicators: Vec<IndicatorValue>,
    pub series: HashMap<String, Vec<SeriesPoint>>,
    /// Administrators only; `None` for cooperative callers.
    pub market_share: Option<MarketShare>,
    pub demographics: Demographics,
    /// Administrators only; per-cooperative rows for the ranking table.
    pub cooperatives: Vec<CooperativeRow>,
}
